Import monitors from sb2 files.

paulkaplan opened the pull request:
Paired with @kchadha on all of this.

### Resolves

_What Github issue does this resolve (please include link)?_

Fixes https://github.com/LLK/scratch-vm/issues/343
Fixes https://github.com/LLK/scratch-vm/issues/790
Fixes https://github.com/LLK/scratch-vm/issues/683
Fixes https://github.com/LLK/scratch-gui/issues/1848

### Proposed Changes

_Describe what this Pull Request does_

Add a step, deferred until after targets are created, where the runtime processes monitors from sb2 files.

Also make the sprite-specific nature of certain monitors more explicit and consistent. 

Also support the fixed monitor ids for sprite specific monitors, etc.

### Reason for Changes

_Explain why these changes should be made_

Monitors! We need them.

### Test Coverage

_Please show how you have added tests to cover your changes_

Added a kind of "kitchen-sink" sb2 import test with a fixture file that looks like this
![image](https://user-images.githubusercontent.com/654102/39817449-2dbadbd8-536c-11e8-98d3-9014fcefac5d.png)

And I added a lot of assertions that cover:
- monitor types set correctly (all 4)
- monitor visibility set correctly
- monitor threads not running if not visible
- monitor positioning imported
- monitor width/height for lists
- custom monitor slider range
- sprite specificity

paulkaplan committed:
Import monitors from sb2 files.

Paired with @kchadha on all of this.

paulkaplan committed:
Fix stage checking

paulkaplan committed:
Fix default value for monitor record

paulkaplan committed:
Remove monitors from sb2 fixtures that expect no threads to stay running

paulkaplan committed:
Add an sb2 import test for importing monitors.

Checks for
- monitor types set correctly (all 4)
- monitor visibility set correctly
- monitor threads not running if not visible
- monitor positioning imported
- monitor width/height for lists
- custom monitor slider range
- sprite specificity

paulkaplan commented:
@kchadha I updated this with fixes for the existing tests and a new fixture/integration test that covers a lot of the edge cases we found.

kchadha commented on src/blocks/scratch3_sensing.js:
```diff
@@ -74,10 +74,18 @@ class Scratch3SensingBlocks {
+                getId: () => 'timer'
+            },
+            sensing_current: {
+                getId: (_, param) => `current_${param}`
```
We should probably add a note here about why this id is different from what's in the default toolbox xml. E.g. we want monitors to be imported as separately from sb2, but we want the block in the toolbox to only ever create one monitor.

I also just realized that this getId function is specific to sb2 now, (e.g. I won't be able to use it in sb3 deserialization) but we can potentially cross that bridge when we get there.

kchadha commented on src/engine/blocks.js:
```diff
@@ -441,22 +441,34 @@ class Blocks {
                     // @todo(vm#565) for numerical values with decimals, some countries use comma
-                    value: ''
+                    value: '',
+                    mode: block.opcode === 'data_listcontents' ? 'list' : 'default'
```
yay for making mode a string instead of a number! =]

kchadha commented on src/serialization/sb2.js:
```diff
@@ -208,6 +209,104 @@ const globalBroadcastMsgStateGenerator = (function () {
 
+/**
+ * Parse a single monitor object and create all its in-memory VM objects.
+ * @param {!object} object - From-JSON "Scratch object"
```
We can provide more info here. The `object` param should represent a monitor and not a sprite (which also qualifies as a 'Scratch object').

kchadha commented on src/serialization/sb2.js:
```diff
@@ -208,6 +209,104 @@ const globalBroadcastMsgStateGenerator = (function () {
+        }
+    }
+
+    // Create a block for the monitor blocks container
```
This comment doesn't correspond with the lines below it.

kchadha commented on src/serialization/sb2.js:
```diff
@@ -208,6 +209,104 @@ const globalBroadcastMsgStateGenerator = (function () {
+    let isSpriteLocalVariable;
+    if (object.cmd === 'getVar:' || object.cmd === 'contentsOfList:') {
+        // These monitors are sprite-specific if they are not targetting the stage.
+        isSpriteLocalVariable = !target.isStage;
```
Just realized... at this point, don't we know that any monitor is sprite specific if it's `target` property is not `'Stage'`? We don't need to do a check for local variables/lists (we normalized lists above).

paulkaplan replied:
Do you mean we could replace the "isSpriteSpecific" logic with just a `!target.isStage`?

kchadha commented on src/serialization/sb2.js:
```diff
@@ -208,6 +209,104 @@ const globalBroadcastMsgStateGenerator = (function () {
+        block.id = getVariableId(object.param);
+    }
+
+    block.id = runtime.monitorBlockInfo.hasOwnProperty(block.opcode) ?
```
I think there should be a comment here which explains that `runtime.monitorBlockInfo` has all monitored block opcodes except `data_variable` and `data_listcontents`, so rewriting the id here doesn't rewrite the id we just set in the block above.

kchadha commented on src/serialization/sb2.js:
```diff
@@ -208,6 +209,104 @@ const globalBroadcastMsgStateGenerator = (function () {
+    // Consult the monitorBlockInfo in the runtime for sprite-specificity.
+    const isSpriteSpecific = isSpriteLocalVariable ||
+        (runtime.monitorBlockInfo.hasOwnProperty(block.opcode) &&
+        runtime.monitorBlockInfo[block.opcode].isSpriteSpecific);
```
we make the `runtime.monitorBlockInfo.hasOwnProperty(block.opcode)` at least twice now, it probably makes sense to make that check and pull out the monitorBlockInfo into a variable up above

paulkaplan committed:
Add comment about sensing block id oddness

kchadha commented on test/integration/monitors.js:
```diff
@@ -0,0 +1,95 @@
+const projectUri = path.resolve(__dirname, '../fixtures/monitors.sb2');
+const project = readFileToBuffer(projectUri);
+
+test('complex', t => {
```
Should this test be called something more descriptive?

paulkaplan committed:
Fix lint error

paulkaplan committed:
Address comments from @kchada review

paulkaplan commented:
@kchadha thanks for the review, updated to all the comments. Did some extra refactoring of the `isSpriteSpecific` area, should make it simpler to read.

paulkaplan committed:
Clear out monitor state on vm.clear

Fixes https://github.com/LLK/scratch-gui/issues/1848

paulkaplan committed:
Do not stringify reported monitor values.

Consulted @fsih, should be fine since formatting is being handled in the GUI anyway.

paulkaplan committed:
Fix clearing monitor state

paulkaplan committed:
Return lists as arrays if running for monitors.

This is a strange quirk of the way the list blocks work, they get joined into a string for interoperability within other other scratch blocks. But for monitors, we need it to stay an array.

kchadha reviewed (approved):
Changes look good to me!

paulkaplan committed:
Return a new copy of the list for monitors when changes occur
